File: 2018/python/day03.py
import re

import numpy as np
from aocd import get_data, submit

DAY = 3
YEAR = 2018

RE_PARSE = re.compile(r"#\d+ @ (\d+),(\d+): (\d+)x(\d+)")


def main() -> tuple[int, int]:
    data = get_data(day=DAY, year=YEAR)
    rectangles = tuple(map(lambda x: tuple(map(int, x)), RE_PARSE.findall(data)))
    grid = np.zeros((1000, 1000), dtype=np.uint)
    for r in rectangles:
        x, y, dx, dy = r
        grid[x : x + dx, y : y + dy] += 1

    part1 = int(np.sum(grid > 1))

    part2 = 0
    for i, r in enumerate(rectangles, 1):
        x, y, dx, dy = r
        if np.all(grid[x : x + dx, y : y + dy] == 1):
            part2 = i
            break

    return part1, part2


if __name__ == "__main__":
    ans1, ans2 = main()
    submit(ans1, part="a", day=DAY, year=YEAR)
    submit(ans2, part="b", day=DAY, year=YEAR)

File: 2018/rust/src/bin/day03.rs
use aocd::*;
use itertools::enumerate;
use ndarray::{s, Array2, ArrayBase, Dim, OwnedRepr};
use regex::Regex;

#[aocd(2018, 3)]
pub fn main() {
    let input = input!();
    let rectangles: Vec<[usize; 4]> = parse_input(input);
    let grid = make_grid(&rectangles);
    submit!(1, part_1(&grid));
    submit!(2, part_2(&grid, &rectangles));
}

fn parse_input(input: String) -> Vec<[usize; 4]> {
    let re = Regex::new(r"#\d+ @ (\d+),(\d+): (\d+)x(\d+)").unwrap();
    re.captures_iter(&input)
        .map(|m| {
            m.iter()
                .skip(1)
                .map(|r| r.unwrap().as_str().parse::<usize>().unwrap())
                .collect::<Vec<usize>>()
                .try_into()
                .unwrap()
        })
        .collect()
}

fn make_grid(rectangles: &Vec<[usize; 4]>) -> ArrayBase<OwnedRepr<i32>, Dim<[usize; 2]>> {
    let mut grid = Array2::zeros((1000, 1000));
    for r in rectangles {
        grid.slice_mut(s![r[0]..r[0] + r[2], r[1]..r[1] + r[3]])
            .scaled_add(1, &Array2::ones((r[2], r[3])))
    }
    grid
}

pub fn part_1(grid: &ArrayBase<OwnedRepr<i32>, Dim<[usize; 2]>>) -> usize {
    grid.fold(0, |mut acc, c| {
        if *c > 1 {
            acc += 1
        };
        acc
    })
}

pub fn part_2(
    grid: &ArrayBase<OwnedRepr<i32>, Dim<[usize; 2]>>,
    rectangles: &Vec<[usize; 4]>,
) -> usize {
    'r: for (i, r) in enumerate(rectangles) {
        for x in grid.slice(s![r[0]..r[0] + r[2], r[1]..r[1] + r[3]]).iter() {
            if *x > 1 {
                continue 'r;
            }
        }
        return i + 1;
    }
    panic!("No valid rectangle found.")
}
